### PR TITLE
Docs for path vars, template functions, and artifact/deploy variables

### PR DESCRIPTION
**Pull request by mitchellh**
This adds complete documentation for:

  * `path` vars
  * `template*` functions
  * Full function reference (completing Evan's work!)
  * `artifact` and `deploy` variables

This **does not** include docs for dynamic config since that is not merged yet to main and also is huge on its own so I want to PR those docs separately.

### CODE REVIEW

File: website/content/docs/waypoint-hcl/functions/template.mdx [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+function such as `templatedir` is called). Custom variables can also
+be provided to the template.
+
+## `templatefile`
+
+```hcl
+templatefile(path[, vars...]) path
+```
+
+`templatefile` takes a path to a file and optionally one more maps of variables
```
Comment (briancain) [inline website/content/docs/waypoint-hcl/functions/template.mdx:28]:
```suggestion
`templatefile` takes a path to a file and optionally one or more maps of variables
```

Not sure if this is what you meant here

**Review by briancain — APPROVED** [review]
Looks good to me!

**Review by evanphx — APPROVED** [review]

File: website/content/docs/waypoint-hcl/variables/artifact.mdx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      }
+    }
+  }
+}
+```
+
+The example uses the `artifact.image` value which is the image name
+(without a tag). Other fields such as `artifact.tag` and
+`artifact.name` (the full name with the tag) are also available.
+For the full list of available field, see the Docker plugin documentation.
```
Comment (krantzinator) [inline website/content/docs/waypoint-hcl/variables/artifact.mdx:54]:
```suggestion
For the full list of available fields, see the Docker plugin documentation.
```

File: website/content/docs/waypoint-hcl/functions/all.mdx [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+# Full Function Reference
+
+This page contains an automatically generated list of all the available
+functions within the Waypoint configuration file. To navigate this page
+more easily, we recommend using your browser's find feature
+(`Cmd-F` or `Ctrl-F`) along with the "Jump to Section" dropdown above.
+
+-> **Note:** This page is currently missing some functions such as
+the template functions (`templatedir`, etc.) and configuration functions
+(`configdynamic`). We are working to fix this. In the mean time, these
```
Comment (krantzinator) [inline website/content/docs/waypoint-hcl/functions/all.mdx:18]:
```suggestion
(`configdynamic`). We are working to fix this. In the meantime, these
```

File: website/content/docs/waypoint-hcl/variables/index.mdx [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+values together:
+
+```hcl
+dockerfile = "${path.app}/Dockerfile"
+```
+
+## Custom Variables
+
+Waypoint currently only supports predefined variables plus the dynamic
+variables introduced by plugins, such as [`artifact`](#). A future version
```
Comment (krantzinator) [inline website/content/docs/waypoint-hcl/variables/index.mdx:37]:
```suggestion
variables introduced by plugins, such as [`artifact`](artifact). A future version
```

**Review by krantzinator — APPROVED** [review]

**Pull request merged**